warehouse/accounts/services.py

```diff
@@ -167,6 +180,63 @@ def verify_email(self, user_id, email_address):
                 email.verified = True
 
 
+@implementer(IUserTokenService)
+class UserTokenService:
+    def __init__(self, user_service, settings):
+        self.user_service = user_service
+        self.serializer = URLSafeTimedSerializer(
+            settings["password_reset.secret"],
+            salt="password-reset",
+        )
+        self.token_max_age = settings["password_reset.token_max_age"]
+
+    def generate_token(self, user):
+        return self.serializer.dumps({
+            "user.id": str(user.id),
+            "user.last_login": str(user.last_login),
+            "user.password_date": str(user.password_date),
+        })
+
+    def get_user_by_token(self, token):
+        if not token:
+            raise InvalidPasswordResetToken(
+                "Invalid token - No token supplied"
+            )
+
+        try:
+            data = self.serializer.loads(token, max_age=self.token_max_age)
+        except SignatureExpired:
+            raise InvalidPasswordResetToken(
+                "Expired token - Token is expired, request a new password "
+                "reset link"
+            )
+        except BadData: #  Catch all other exceptions
+            raise InvalidPasswordResetToken(
+                "Invalid token - Request a new password reset link"
+            )
+
+        # Check whether a user with the given user ID exists
+        user = self.user_service.get_user(uuid.UUID(data.get("user.id")))
+        if user is None:
+            raise InvalidPasswordResetToken("Invalid token - User not found")
+
+        last_login = data.get("user.last_login")
+        if str(user.last_login) > last_login:
+            raise InvalidPasswordResetToken(
+                "Invalid token - User has logged in since this token was "
+                "requested"
+            )  # TODO: track and audit this, seems alertable
+
+        password_date = data.get("user.password_date")
+        if str(user.password_date) > password_date:
+            raise InvalidPasswordResetToken(
+                "Invalid token - Another password reset has been completed "
```

Comment:
Slight tweak here, we might want to just say that the password has changed since the token was requested (or use an OR statement). If they're already logged in and manually change their password they'll hit this case too.